sw/device/lib/dif/dif_plic.h

```diff
@@ -0,0 +1,248 @@
+ *
+ * Enumeration used to determine which peripheral asserted the corresponding
+ * IRQ source.
+ */
+typedef enum dif_plic_irq_peripheral {
+  kDifPlicPeripheralNone = 0, /**< none */
+  kDifPlicGpio,               /**< GPIO */
+  kDifPlicUart,               /**< UART */
+  kDifPlicSpiDevice,          /**< SPI */
+  kDifPlicFlashCtrl,          /**< Flash */
+  kDifPlicHmac,               /**< HMAC */
+  kDifPlicAlertHandler,       /**< Alert handler */
+  kDifPlicNmiGen,             /**< NMI generator */
+} dif_plic_irq_peripheral_t;
+
+/**
+ * PLIC external interrupt target
+ *
+ * Enumeration used to determine which set of IE, CC, threshold registers to
+ * access dependent on the target.
+ */
+typedef enum dif_plic_target {
+  kDifPlicTargetNone = 0,
+  kDifPlicIbex0,
+} dif_plic_target_t;
+
+/**
+ * Generic enable/disable enumeration
+ *
+ * Enumeration used to enable/disable bits, flags, ...
+ */
+typedef enum dif_plic_on_off {
+  kDifPlicEnable = 0, /**< enable */
+  kDifPlicDisable,    /**< disable */
+} dif_plic_on_off_t;
+
+/**
+ * PLIC CC (IRQ claim register data)
+ *
+ * Data that is populated by the PLIC DIF when interrupt is claimed. Can be
+ * used by the PLIC DIF consumers to establish which peripheral issued an IRQ,
+ * and the type of this IRQ.
+ */
+typedef struct dif_irq_data {
+  dif_plic_irq_peripheral_t peripheral; /**< Peripheral that interrupted */
+  dif_plic_irq_src_id_t source;         /**< IRQ type */
+} dif_irq_data_t;
+
+/**
+ * PLIC instance state
+ *
+ * PLIC persistent data that is required by all the PLIC API. The PLIC DIF
+ * consumers shall not modify the |claimed_irq|, and are only allowed to read
+ * (examine) the contents. |claimed_irq| is only set in the dif_plic_claim_irq
+ * function, which then must be cleared with a call to
+ * dif_plic_clear_claimed_irq when the caller has finished processing the IRQ.
+ *
+ * |target| dictates which target specific registers IE, CC, THRESHOLD shall be
+ * used. IP, LE and PRIO registers are not target specific and are not affected
+ * by this value.
+ */
+typedef struct dif_plic {
+  reg32_t base_addr;          /**< PLIC instance base address */
+  dif_plic_target_t target;   /**< PLIC target */
+  dif_irq_data_t claimed_irq; /**< PLIC IRQ data */
+} dif_plic_t;
+
+/**
+ * Initialise an instance of PLIC
+ *
+ * Initialise a PLIC instance. Information that must be retained, and is
+ * required to program PLIC, shall be stored in @p arg3. PLIC supports external
+ * interrupt delivery to multiple targets, which are requested in @p arg2.
+ *
+ * Note that only target specific registers are initialised, the PLIC target
+ * common registers must be initialised by the caller.
+ * @param base_addr Base address of an instance of the PLIC IP block
+ * @param target Target that is interrupted by PLIC.
+ * @param plic PLIC state data
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_plic_init(uintptr_t base_addr, dif_plic_target_t target,
```

Comment:
Can you make this accept a reg32_t, instead? Under test, it might not be possible to construct a reg32_t from an address, and reg32_t is supposed to be an encapsulation boundary.

Reply:
Not applicable anymore, target and claim_data is no longer part of the dif_plic state